src/sgame/sg_admin.cpp

```diff
@@ -6077,18 +6076,19 @@ bool G_admin_bot( gentity_t *ent )
 bool G_admin_listbots( gentity_t *ent )
 {
 	ADMP( va( "%s %d", QQ( N_( "^3listbots:^* $1$ bots in game:") ), level.numPlayingBots ) );
-	ADMP( QQ( N_( "Slot Name Team [s=skill b=behavior g=goal]" ) ) );
-	ADMBP_begin();
-	ForEntities<ClientComponent>( []( Entity& entity, ClientComponent& )
+	ADMP( QQ( N_( "Slot Name Team [s=skill ss=skillset b=behavior g=goal]" ) ) );
+	ForEntities<ClientComponent>( [ent]( Entity& entity, ClientComponent& )
 	{
-		gentity_t* ent = entity.oldEnt;
-		if ( !( ent->r.svFlags & SVF_BOT ) )
+		gentity_t* bot = entity.oldEnt;
+		if ( !( bot->r.svFlags & SVF_BOT ) )
 		{
 			return;
 		}
-		ADMBP( va( "%zu %s", ent - g_entities, G_BotToString( ent ).c_str() ) );
+		AP( Str::Format( "print_tr %s %d %s",
```

Comment:
given no translation will be done, no need to use `print_tr`

Reply:
If I get this correctly, You will be sending the listbots text and the table header using ADMP, which hardcodes the `ent` variable and only sends it to the `ent` player. However, the actual contents of list command, you will send each line via `AP` (ie, to **all** players). this seems undesirable. Just use trap_SendServerCommand at this rate...

Given that this is a debugging command, I'm not super concerned about the ServerCommand spam.

Reply:
I mean I would be using ADMP if I could get it to work. But it just doesn't want to work. I'd be grateful if you could figure that one out for me :-/

I think ADMP is now my #1 hated function.

Reply:
ADMP is just a hacky way to batch sending large swaths of strings to the client. Your whole mechanism for avoiding overflow depends on **not** batching anything and sending everything. This is obviously inefficient, but required due to the possibly large amounts of text being sent per line.

Instead you want to replace your AP call with

```
trap_SendServerCommand( ent->num(), Str::Format( blah ).c_str() );
```

`AP(arg)` is basically a define for `trap_SendServerCommand( -1, arg );`, and `AP` stands for "All Players" whereas I think `ADMBP` stands for "ADMin Buffered Print".